### DISCUSSION TITLE
Validating a field's value is unique across yaml files?

### DISCUSSION THREAD

**Discussion by deepbrook** [cue-lang/cue#2728] [Q&A] [answered]
Ahoi!

I have the following schema:

```
 #Config: {
    name!: string
    prefix!: string
#Config
```

And two yaml files as follows:

```shell
> cat alice.yaml
---
name: alice
prefix: great
> cat bob.yaml
---
name: bob
prefix: rounded
> cat charlie.yaml
---
name: charlie
prefix: great
```

If want to ensure that the `prefix` field's value is unique across all yaml files in a directory (or alternatively, within the same schema - if it's possible to use `cue import -l 'name'` or something to merge them?).

Effectively, I want something like `cue cmd validate charlie.yaml --against alice.yaml,bob.yaml`, to validate `prefix` in the file `charlie.yaml` as if it have a validation pattern like  this `prefix!: [!=~"(great|great).*"]`

Is this possible with cue? Thanks for the help!

**Comment by verdverm** [comment, marked as answer]
I think something like the following

1. `cue vet schema.cue *.yaml -l 'data:' -l 'name'` ... this will bring all the CUE and Yaml together, the first part places it under a known field we can use later, the second part places each file under a field in data
2. apply the schema with `data: [string]: #Config` ... the pattern constraint here should match your `-l/--path` flag usage
3. use an extra field and comprehension to enforce the uniqueness `unique: true & list.UniqueItems([ for d in data { d.prefix }])` (https://pkg.go.dev/cuelang.org/go@v0.7.0/pkg/list#UniqueItems)

**Reply by deepbrook** [reply]
```
 import (
	"list"
)
 #Config: {
    name!: string
    prefix!: string
}
 data: [string]: #Config
 prefix_unique: true & list.UniqueItems([ for d in data { d.prefix }])
 ```
 
 Validates as described using `vet`:
 
 ```shell
🐟 ❯ cue vet *.yaml schema.cue --path "data:" --path "name"
prefix_unique: conflicting values false and true:
    ./schema.cue:9:16
    ./schema.cue:9:23
 ```
 
 I'm not sure how to implement this as a command, but my guess is I'll need a package for the schema and then do something like this:
 
```
cmd: validate: {
    data: [string]: #Config
    prefix_unique: true & list.UniqueItems([ for d in data { d.prefix }])
}
```
Might have to pre-compile it first (`cue import`?). 

Anyway, thank you! That pushed me in the right direction. Cheers!
